janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java

```diff
@@ -498,58 +515,134 @@ private static Query numericQuery(String key, Cmp relation, Number value) {
+        } else {
+            final Analyzer analyzer = delegatingAnalyzer.getWrappedAnalyzer(key);
+            if (analyzer != null) {
+                terms = customTokenize(analyzer, key, value);
+            } else {
+                switch (mapping) {
+                    case STRING:
+                        terms = Collections.singletonList(value);
+                        break;
+                    case DEFAULT:
+                    case TEXT:
+                        terms = Text.tokenize(value);
+                        break;
+                    default:
+                        throw new RuntimeException("unsupported case " + mapping);
+                }
+            }
+        }
+
+        if (terms.isEmpty()) {
+            // This might happen with very short terms
+            if (janusgraphPredicate == Text.CONTAINS_PREFIX) {
+                final Term term;
+                if (mapping == Mapping.STRING) {
+                    term = new Term(key, value);
+                } else {
+                    term = new Term(key, value.toLowerCase());
+                }
+                params.addQuery(new PrefixQuery(term), BooleanClause.Occur.MUST);
+            }
+        } else if (terms.size() == 1) {
+            if (janusgraphPredicate == Cmp.EQUAL || janusgraphPredicate == Text.CONTAINS) {
+                params.addQuery(new TermQuery(new Term(key, value)));
+            } else if (janusgraphPredicate == Cmp.NOT_EQUAL) {
+                final BooleanQuery.Builder q = new BooleanQuery.Builder();
+                q.add(new MatchAllDocsQuery(), BooleanClause.Occur.MUST);
+                q.add(new TermQuery(new Term(key, value)), BooleanClause.Occur.MUST_NOT);
+                params.addQuery(q.build(), BooleanClause.Occur.MUST);
+            } else if (janusgraphPredicate == Text.CONTAINS_PREFIX) {
+                final Term term;
+                if (mapping == Mapping.STRING) {
+                    term = new Term(key, terms.get(0));
+                } else {
+                    term = new Term(key, terms.get(0).toLowerCase());
+                }
+                params.addQuery(new PrefixQuery(term), BooleanClause.Occur.MUST);
+            }
+        } else {
+            // at the moment, this is only walked for EQUAL and Text.CONTAINS (String and Text mappings)
+            final BooleanQuery.Builder q = new BooleanQuery.Builder();
+            for (final String term : terms) {
+                q.add(new TermQuery(new Term(key, term)), BooleanClause.Occur.MUST);
+            }
+            params.addQuery(q.build());
+        }
+    }
+
+    private LuceneCustomAnalyzer delegatingAnalyzerFor(String store, KeyInformation.IndexRetriever information2) {
+        LuceneCustomAnalyzer delegatingAnalyzer = delegatingAnalyzers.get(store);
+        if (delegatingAnalyzer == null) {
+            delegatingAnalyzer = new LuceneCustomAnalyzer(store, information2, Analyzer.PER_FIELD_REUSE_STRATEGY);
```

Comment:
We do, the LuceneCustomAnalyzer lives to dispatch different fields to different analyzers. 
This is also why we need a LuceneCustomAnalyzer for each store.
So analyzers are cached first by store name (in delegatingAnalyzers map) and then by field name (in the analyzer map of the corresponding LuceneCustomAnalyzer)

I wouldn't be surprised if there was a lucene way to turn all stores into 1 lucene index (with 1 indexWriter delegating writes of each store to a corresponding sub-IndexWriter, same things for reads) but LuceneIndex wasn't written like this in the first place (which is why we have a map to IndexWriters : LuceneIndex does it it's own way, I guess)

Reply:
Ok. This is what I think :

In the original LuceneIndex implementation, 
a StandardAnalyzer is used to tokenize TEXT strings (stored in a TextField) but STRING strings are stored as is in a StringField. The an IndexWriter writes that to disk

When Querying, LuceneIndex uses Text.tokenize() in Text.CONTAINS, Text.CONTAINS_FUZZY querries
which is really really wrong because strings tokenized through an analyzer should be querried through the same Analyzer (or you take the risk to miss everything if the tokens produced aren't the same)

LuceneIndex does quite a few other evils, and tests don't look like they cover all corners cases.

When I first hacked custom analyzer support into this. I tried to reuse what LuceneIndex did, in order to make things work (I still only have a shallow understanding of what Janusgraph does sometimes and I have some gripes with the way it does things sometimes...like index mutation).

Now, Custom Analyzer support requires TextField and no StringField (or your custom analyzer for mapping string won't work, because lucene won't even use it with StringField). So, a correct and clean solution should probably remove all that Text.tokenize stuff and make sure all produced tokens are the same when writing and reading them.

-> we need to make everything pass through an Analyzer (Standard for text and Keyword for string)...

Reply:
At the moment, all tests pass with my modifications. 
But the convertQuery routine should definitely be rewritten to improve what is done in there.

After this pr, I was planning to do that and send more pr for more tests, bug fixes, better code and more lucene features next (for example, at the moment, LuceneIndex loses data when a mutation happens)